handler/handler.go

```diff
@@ -4,20 +4,67 @@ import (
 	"encoding/json"
 	"fmt"
 	"log"
+	"sync"
+	"time"
 
 	"github.com/nats-io/go-nats-streaming"
 	"github.com/openfaas/faas/gateway/queue"
 )
 
 // NatsQueue queue for work
 type NatsQueue struct {
-	nc        stan.Conn
+	nc             stan.Conn
+	ncMutex        *sync.RWMutex
+	maxReconnect   int
+	reconnectDelay time.Duration
+
 	ClientID  string
 	ClusterID string
 	NATSURL   string
 	Topic     string
 }
 
+func (q *NatsQueue) connect() error {
+	nc, err := stan.Connect(
+		q.ClusterID,
+		q.ClientID,
+		stan.NatsURL(q.NATSURL),
+		stan.SetConnectionLostHandler(func(conn stan.Conn, err error) {
+			log.Printf("Disconnected from %s\n", q.NATSURL)
+
+			q.reconnect()
+		}),
+	)
+
+	if err != nil {
+		return err
+	}
+
+	q.ncMutex.Lock()
+	q.nc = nc
+	q.ncMutex.Unlock()
+
+	return nil
+}
+
+func (q *NatsQueue) reconnect() {
+	for i := 0; i < q.maxReconnect; i++ {
+		func(i int) {
```

Comment:
The comment from Ivan was about using a `select`, `cancel` and shutdown mechanism which we don't have in the code right now so this change isn't needed. Please can you revert to the loop with `time.Sleep` and re-test so that I can merge into master?

FYI: [This is an example of using select and a time.After](https://github.com/alexellis/inlets/blob/sync/pkg/server/server.go#L106)

Reply:
Yup, I know, I wanted to implemented it now to make it easier to add shutdown logic later, but reverted the change